src/libraries/System.Private.CoreLib/src/System/Globalization/CalendarData.Icu.cs

```diff
@@ -422,15 +422,29 @@ internal static bool EnumCalendarInfo(string localeName, CalendarId calendarId,
             return result;
         }
 
+#if !TARGET_BROWSER
         private static unsafe bool EnumCalendarInfo(string localeName, CalendarId calendarId, CalendarDataType dataType, ref IcuEnumCalendarsData callbackContext)
         {
             return Interop.Globalization.EnumCalendarInfo(EnumCalendarInfoCallback, localeName, calendarId, dataType, (IntPtr)Unsafe.AsPointer(ref callbackContext));
         }
+#else
+        private static unsafe bool EnumCalendarInfo(string localeName, CalendarId calendarId, CalendarDataType dataType, ref IcuEnumCalendarsData callbackContext)
+        {
+            // Temp workaround for pinvoke callbacks for Mono-Wasm-Interpreter
+            // https://github.com/dotnet/runtime/issues/39100
+            var calendarInfoCallback = new Interop.Globalization.EnumCalendarInfoCallback(EnumCalendarInfoCallback);
+            return Interop.Globalization.EnumCalendarInfo(
+                System.Runtime.InteropServices.Marshal.GetFunctionPointerForDelegate(calendarInfoCallback),
+                localeName, calendarId, dataType, (IntPtr)Unsafe.AsPointer(ref callbackContext));
+        }
 
-        private static unsafe void EnumCalendarInfoCallback(string calendarString, IntPtr context)
+        [Mono.MonoPInvokeCallback(typeof(Interop.Globalization.EnumCalendarInfoCallback))]
+#endif
+        private static unsafe void EnumCalendarInfoCallback(char* calendarStringPtr, IntPtr context)
         {
             try
             {
+                var calendarString = new string(calendarStringPtr);
```

Comment:
>calendarString [](start = 20, length = 14)

is it worth it to delay creating the string till the line we add it to the results? and before that we can use Span?

Reply:
@tarekgh it's also used inside the loop so I don't see how moving it anywhere can help

Reply:
inside the loop you will use the compare method using the span [https://github.com/tarekgh/runtime/blob/38f6ebca112d838b4dbfc8fe32cff1b4c7275ffd/src/libraries/System.Private.CoreLib/src/System/String.Comparison.cs#L433](https://github.com/tarekgh/runtime/blob/38f6ebca112d838b4dbfc8fe32cff1b4c7275ffd/src/libraries/System.Private.CoreLib/src/System/String.Comparison.cs#L433)

---
In reply to: [455429446](https://github.com/dotnet/runtime/pull/37971#discussion_r455429446) [](ancestors = 455429446)

Reply:
this is not that important to do if you think will complicate the code as the case which we'll optimize is not main scenario cases. treat this as nit

---
In reply to: [455430821](https://github.com/dotnet/runtime/pull/37971#discussion_r455430821) [](ancestors = 455430821,455429446)

Reply:
changed to Span, thanks